File: src/Medium/DynamicTest/maxProfit.java
package Medium.DynamicTest;



/**
 * 309. 最佳买卖股票时机含冷冻期
 * 给定一个整数数组，其中第 i 个元素代表了第 i 天的股票价格 。​
 *
 * 设计一个算法计算出最大利润。在满足以下约束条件下，你可以尽可能地完成更多的交易（多次买卖一支股票）:
 *
 * 你不能同时参与多笔交易（你必须在再次购买前出售掉之前的股票）。
 * 卖出股票后，你无法在第二天买入股票 (即冷冻期为 1 天)。
 * 示例:
 *
 * 输入: [1,2,3,0,2]
 * 输出: 3
 * 解释: 对应的交易状态为: [买入, 卖出, 冷冻期, 买入, 卖出]*/

/**
 * @author 马世臣
 * @// TODO: 2020/3/14  */

public class maxProfit {




    public int maxProfit(int[] prices) {
        int len;
        if (prices == null || (len = prices.length) == 0)
            return 0;
        int[][] dp = new int[len][3];
        dp[0][0] = 0; // 不持股
        dp[0][1] = -prices[0]; // 持股
        dp[0][2] = 0; // 冷冻期
        for (int i = 1; i < len; i++) {
            // 不持股
            // 第i-1天不持股，第i天仍不持股
            // 第i-1天处于冷冻期，第i天就变为不持股状态，那么当前第i天的最大值就是从这两个状态中取
            dp[i][0] = Math.max(dp[i - 1][0], dp[i - 1][2]);
            // 持股
            // 第i-1天持股，第i天仍持股
            // 第i-1天不持股，第i天买入
            dp[i][1] = Math.max(dp[i - 1][1], dp[i - 1][0] - prices[i]);
            // 冷冻期
            // 第i-1天持有股票且今天卖出
            dp[i][2] = dp[i - 1][1] + prices[i];
        }
        // 返回不持由股票或者处于冷冻期的最大值
        return Math.max(dp[len - 1][0], dp[len - 1][2]);
    }

    public static void main(String[] args) {

    }
}

File: src/Difficult/DynamicTest/maxProfit.py
import time, re
from typing import List, Tuple, Union, Optional
from heapq import heappop, heappush, heapify, heappushpop, heapreplace
from collections import defaultdict, deque, Counter
from itertools import accumulate, permutations, combinations, product, compress, zip_longest, pairwise, groupby
from math import perm, comb, gcd, lcm, inf, ceil, floor, factorial, dist, sqrt
from functools import cache, lru_cache, reduce
from sortedcontainers import SortedList, SortedSet, SortedDict
from bisect import bisect_left, bisect_right, insort, insort_left, insort_right


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        n = len(prices)
        dpl = [0] * n
        preMin = prices[0]
        for i in range(1, n):
            dpl[i] = max(dpl[i - 1], prices[i] - preMin)
            preMin = min(preMin, prices[i])

        dpr = [0] * n
        preMax = prices[-1]
        for i in range(n - 2, -1, -1):
            dpr[i] = max(dpr[i + 1], preMax - prices[i])
            preMax = max(preMax, prices[i])

        res = 0
        for i in range(n - 1):
            res = max(res, dpl[i] + dpr[i + 1])
        res = max(res, dpl[-1], dpr[0])
        return res

    def maxProfit2(self, k: int, prices: List[int]) -> int:
        # @cache
        # def single(i, j):
        #     if i > j:
        #         return 0
        #     preMin = prices[i]
        #     res = 0
        #     for k in range(i + 1, j + 1):
        #         res = max(res, prices[k] - preMin)
        #         preMin = min(prices[k], preMin)
        #     return res
        #
        # @cache
        # def f(e, k):
        #     if k == 1:
        #         return single(0, e)
        #     res = 0
        #     for i in range(e + 1):
        #         res = max(res, f(i, k - 1) + single(i + 1, e))
        #     return res
        #
        # return f(len(prices) - 1, k)
        dp = [[-inf, 0] for _ in range(k)]
        for p in prices:
            dp[0][0] = max(dp[0][0], -p)
            dp[0][1] = max(dp[0][1], dp[0][0] + p)
            for i in range(1, k):
                dp[i][0] = max(dp[i][0], dp[i - 1][1] - p)
                dp[i][1] = max(dp[i][1], dp[i][0] + p)

        return dp[k - 1][1]


if __name__ == '__main__':
    s = Solution()
    # print(s.maxProfit(prices=[3, 3, 5, 0, 0, 3, 1, 4]))
    # print(s.maxProfit(prices=[1, 2, 3, 4, 5]))
    print(s.maxProfit2(k=2, prices=[2, 4, 1]))
    print(s.maxProfit2(k=2, prices=[3, 2, 6, 5, 0, 3]))
